File: code/ch13/13.1.2.best_time_to_buy_and_sell_stock.py
from typing import List
import sys


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        max_diff = 0
        min_price = sys.maxsize
        for i in range(len(prices)):
            if min_price > prices[i]:
                min_price = prices[i]
            max_diff = max(max_diff, prices[i] - min_price)
        return max_diff

File: code/ch13/13.1.2.best_time_to_buy_and_sell_stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if (prices.size() == 0)
            return 0;

        int max_diff = 0;
        int min_price = prices[0];
        
        for (int i = 1; i < prices.size(); i++) {
            if (min_price > prices[i])
                min_price = prices[i];
            max_diff = max(max_diff, prices[i] - min_price);
        }
        
        return max_diff;
    }
};
